### PR TITLE
HostConfig for /start endpoint

### PR DESCRIPTION
**Pull request by vieux**
replace https://github.com/docker/swarm/pull/1829/
fix https://github.com/docker/swarm/issues/1781

### CODE REVIEW

**Comment by nishanttotla** [comment]
I'm a little unsure about when to make this change, since it certainly conflicts with using container config structs from `engine-api`.

**Comment by vieux** [comment]
@nishanttotla it fixes an issue in 1.1 I don't see why not merge it

File: api/handlers.go [hunk 1]
```diff
@@ -749,7 +749,28 @@ func postContainersStart(c *context, w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	buf, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		httpError(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	r.Body.Close()
+
+	fmt.Println(len(buf))
```
Comment (nishanttotla) [inline api/handlers.go]: @vieux did you mean to remove this print statement?

Comment (vieux) [inline api/handlers.go]: yes good catch, thanks

Comment (vieux) [inline api/handlers.go]: fixed

**Comment by vieux** [comment]
ping @dongluochen

**Comment by nishanttotla** [comment]
@vieux I thought about it some more, and the reason I was unsure was that `engine-api` doesn't allow for a `HostConfig` on container start. So this change may not be as simple eventually: https://github.com/docker/engine-api/blob/master/client/container_start.go

**Comment by vieux** [comment]
@nishanttotla you'll have to change engine-api then, as long as our minimum supported engine version ca do it, we should do it :)

**Comment by nishanttotla** [comment]
Aaah it's so complicated :)

It seems like the daemon code still supports this: https://github.com/docker/docker/blob/master/daemon/start.go#L16

But the plan is to phase it out in the upcoming release, so I really doubt we can build it into `engine-api`, but I'll ask around.

File: api/handlers.go [hunk 2]
```diff
@@ -749,7 +749,27 @@ func postContainersStart(c *context, w http.ResponseWriter, r *http.Request) {
...
+	hostConfig := &dockerclient.HostConfig{
+		MemorySwappiness: -1,
+	}
+
+	buf, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		httpError(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	r.Body.Close()
```
Comment (dongluochen) [inline api/handlers.go]: My understanding is that this is unnecessary. Request body should be close by the transport automatically.

**Comment by dongluochen** [comment]
I think it's a complicated issue in the middle migrating client. I'm fine with either decision.

**Comment by nishanttotla** [comment]
Even if we upped the minimum supported Engine version, we will have to support this for a little while at least.

One suggestion I have is to update config of the container first if a `HostConfig` is passed on start using `ContainerUpdate` (https://github.com/docker/engine-api/blob/master/client/container_update.go). Then call normal start without `HostConfig`. Is that a reasonable solution?

**Comment by vieux** [comment]
@nishanttotla it might, but let's not block this PR it can be resolved after.

**Comment by nishanttotla** [comment]
LGTM

**Comment by dongluochen** [comment]
LGTM besides the line with `r.Body.Close()`.

**Comment by nishanttotla** [comment]
cc @calavera

File: api/handlers.go (hunk 2 above)
Comment (calavera) [inline api/handlers.go]: Closing doesn't have any side effect, so it should be okay to close it here. You cannot read the buffer again anyways.

**Comment by dongluochen** [comment]
LGTM

**Pull request merged**